lib/models/application.coffee

```diff
@@ -202,35 +202,37 @@ getApplicationModel = (deps, opts) ->
 	# 	console.log(application);
 	# });
 	###
-	exports.get = (nameOrId, options = {}, callback) ->
+	exports.get = (nameOrSlugOrId, options = {}, callback) ->
 		callback = findCallback(arguments)
 
 		Promise.try ->
-			if not nameOrId?
-				throw new errors.BalenaApplicationNotFound(nameOrId)
+			if not nameOrSlugOrId?
+				throw new errors.BalenaApplicationNotFound(nameOrSlugOrId)
 
-			if isId(nameOrId)
+			if isId(nameOrSlugOrId)
 				pine.get
 					resource: 'application'
-					id: nameOrId
+					id: nameOrSlugOrId
 					options: mergePineOptions({}, options)
 				.tap (application) ->
 					if not application?
-						throw new errors.BalenaApplicationNotFound(nameOrId)
+						throw new errors.BalenaApplicationNotFound(nameOrSlugOrId)
 			else
 				pine.get
 					resource: 'application'
 					options:
 						mergePineOptions
 							$filter:
-								app_name: nameOrId
+								$or:
+									app_name: nameOrSlugOrId
+									slug: nameOrSlugOrId
```

Comment:
I think it's worth converting to lowercase:

```suggestion
									slug: nameOrSlugOrId.toLowerCase()
```

Reply:
If it was true (is it?!) that a slug always includes a slash character (`'/'`) and an app name never does, then the SDK could decide which query to make instead of asking the database to search for both (i.e., avoiding the need for `$or`).

Reply:
I'm not sure about that :thinking: 
Imo it feels weird thinking that a user has an app named `MyApp` and doing `username/MyaPP` still gives back results.
Moreover, we already do not do such a conversion for the app name, even though that's also unique per user on its lower case form. So doing it for the slug feels a bit inconsistent.
See: https://github.com/balena-io/balena-api/blob/51f6c5e67553d6fd57824fd720f2d96ac3c609ad/src/models/hooks/resources/application.ts#L196-L218

I think we should leave the casing changing for SDK consumers, eg the CLI.
An SDK consumer might be indeed be interested on getting an app by exact slug casing.
Another, not that related though is that the git url (which is tightly coupled with the slug) is case sensitive, so doing the same here kind of makes sense to me.

Reply:
@pdcastro the slug always has a `/` on balenaCloud, but on open balena the slug is just the `app_name.toLowerCase()` (since there are no owners). We could still have a check to not filter on the `app_name` if there is a `/`, but that still doesn't allow use to be able to filter only on a single field for all cases and environments.

If we expect that's going to have measurable impact on the DBs load, then it makes sense to add it and avoid the $or iff the arg contains a `/`. On the other hand this will add a differentiation point based on a check that only exists on a hook and isn't part of the model.

I'm partially convinced of adding that though. What do you think?

Reply:
`application.get` is a very common operation, but I don't know if the impact on DB/API load would be measurable. The fact that openBalena slugs don't include a `'/'` spoils my suggestion, and makes the extra logic harder to justify (thinking of Petro's summit presentation on avoiding `'if's`). It's your call -- I'm happy for the PR to be merged as it is. :-)

Reply:
I still think converting to lowercase is best. This is because the API implicitly manages the slug for you behind the scenes. The SDK converting to lower case is the flip side of that coin. Ideally the model itself would automatically do case-insensitive comparisons against that field, but having the SDK normalise the input value is the next best thing.

Reply:
@dfunckt @sradevski It still feels weird to me to be able to find an app by using `THGREASI/TESTSTARTER`, and prefer the strict-over-loose when in doubt, but I see your point.
Pushed the suggested change as a separate commit.
Let's see what @shaunmulligan thinks about this whole PR.